File: __tests__/utils/seasonLogic.test.ts
import {
  maxDayForMonth,
  isOffSeason,
  nextSeasonChangeDate,
  isDegenerate,
  OFFSEASON_START,
  OFFSEASON_END,
} from '../../src/utils/seasonLogic';

// ─── maxDayForMonth ───────────────────────────────────────────────────────────

describe('maxDayForMonth', () => {
  test.each([
    [1, 31], [3, 31], [5, 31], [7, 31], [8, 31], [10, 31], [12, 31],
  ])('mois %i a 31 jours', (month, expected) => {
    expect(maxDayForMonth(month)).toBe(expected);
  });

  test.each([
    [4, 30], [6, 30], [9, 30], [11, 30],
  ])('mois %i a 30 jours', (month, expected) => {
    expect(maxDayForMonth(month)).toBe(expected);
  });

  it('février a 28 jours (pas de gestion bissextile)', () => {
    expect(maxDayForMonth(2)).toBe(28);
  });
});

// ─── isOffSeason — fenêtre par défaut (1er nov → 31 mars) ────────────────────

describe('isOffSeason — fenêtre par défaut', () => {
  it('milieu de l\'hiver (15 déc) → hors-saison', () => {
    expect(isOffSeason(new Date(2026, 11, 15))).toBe(true);
  });

  it('milieu de l\'été (15 juil) → en saison', () => {
    expect(isOffSeason(new Date(2026, 6, 15))).toBe(false);
  });

  it('1er novembre : premier jour hors-saison → true', () => {
    expect(isOffSeason(new Date(2026, 10, 1))).toBe(true);
  });

  it('31 octobre : veille du début → false', () => {
    expect(isOffSeason(new Date(2026, 9, 31))).toBe(false);
  });

  it('31 mars : dernier jour hors-saison → true', () => {
    expect(isOffSeason(new Date(2026, 2, 31))).toBe(true);
  });

  it('1er avril : lendemain de la fin → false', () => {
    expect(isOffSeason(new Date(2026, 3, 1))).toBe(false);
  });

  it('1er janvier (la fenêtre chevauche le 31 déc) → hors-saison', () => {
    expect(isOffSeason(new Date(2026, 0, 1))).toBe(true);
  });

  it('15 décembre → hors-saison (même comportement en changeant d\'année)', () => {
    expect(isOffSeason(new Date(2025, 11, 15))).toBe(true);
    expect(isOffSeason(new Date(2027, 11, 15))).toBe(true);
  });
});

// ─── isOffSeason — fenêtre personnalisée ─────────────────────────────────────

describe('isOffSeason — fenêtre personnalisée', () => {
  const start = {month: 6, day: 1};  // 1er juin
  const end   = {month: 9, day: 30}; // 30 sept (fenêtre dans l'année)

  it('15 août : dans la fenêtre → true', () => {
    expect(isOffSeason(new Date(2026, 7, 15), start, end)).toBe(true);
  });

  it('1er juin : premier jour → true', () => {
    expect(isOffSeason(new Date(2026, 5, 1), start, end)).toBe(true);
  });

  it('31 mai : veille → false', () => {
    expect(isOffSeason(new Date(2026, 4, 31), start, end)).toBe(false);
  });

  it('30 sept : dernier jour → true', () => {
    expect(isOffSeason(new Date(2026, 8, 30), start, end)).toBe(true);
  });

  it('1er octobre : hors fenêtre → false', () => {
    expect(isOffSeason(new Date(2026, 9, 1), start, end)).toBe(false);
  });

  it('15 déc : hors fenêtre → false', () => {
    expect(isOffSeason(new Date(2026, 11, 15), start, end)).toBe(false);
  });
});

// ─── nextSeasonChangeDate ─────────────────────────────────────────────────────

describe('nextSeasonChangeDate', () => {
  it('en hiver → prochaine date = fin hors-saison (31 mars)', () => {
    const result = nextSeasonChangeDate(new Date(2026, 11, 15)); // 15 déc
    expect(result.getMonth() + 1).toBe(OFFSEASON_END.month);  // 3 = mars
    expect(result.getDate()).toBe(OFFSEASON_END.day);           // 31
  });

  it('en été → prochaine date = début hors-saison (1er nov)', () => {
    const result = nextSeasonChangeDate(new Date(2026, 6, 15)); // 15 juil
    expect(result.getMonth() + 1).toBe(OFFSEASON_START.month); // 11 = nov
    expect(result.getDate()).toBe(OFFSEASON_START.day);          // 1
  });

  it('en février → fin hors-saison cette année (mars n\'est pas encore passé)', () => {
    const result = nextSeasonChangeDate(new Date(2026, 1, 15)); // 15 fév
    expect(result.getFullYear()).toBe(2026);
    expect(result.getMonth() + 1).toBe(3); // mars
    expect(result.getDate()).toBe(31);
  });

  it('en juillet → début hors-saison cette année (nov pas encore passé)', () => {
    const result = nextSeasonChangeDate(new Date(2026, 6, 1)); // 1er juil
    expect(result.getFullYear()).toBe(2026);
    expect(result.getMonth() + 1).toBe(11); // nov
  });
});

// ─── isDegenerate ─────────────────────────────────────────────────────────────

describe('isDegenerate', () => {
  it('start === end → dégénéré', () => {
    expect(isDegenerate({month: 6, day: 15}, {month: 6, day: 15})).toBe(true);
  });

  it('start !== end (config par défaut) → non dégénéré', () => {
    expect(isDegenerate(OFFSEASON_START, OFFSEASON_END)).toBe(false);
  });

  it('dates différentes → non dégénéré', () => {
    expect(isDegenerate({month: 11, day: 1}, {month: 3, day: 31})).toBe(false);
  });
});

File: app/src/test/java/com/almostblue/domain/SeasonLogicTest.kt
package com.almostblue.domain

import java.time.LocalDate
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** Port 1:1 de spec/__tests__/utils/seasonLogic.test.ts. */
class SeasonLogicTest {

    // ─── maxDayForMonth ─────────────────────────────────────────────────────────

    @Test
    fun `mois a 31 jours`() {
        for (month in listOf(1, 3, 5, 7, 8, 10, 12)) {
            assertEquals("mois $month", 31, maxDayForMonth(month))
        }
    }

    @Test
    fun `mois a 30 jours`() {
        for (month in listOf(4, 6, 9, 11)) {
            assertEquals("mois $month", 30, maxDayForMonth(month))
        }
    }

    @Test
    fun `fevrier a 28 jours (pas de gestion bissextile)`() {
        assertEquals(28, maxDayForMonth(2))
    }

    // ─── isOffSeason — fenêtre par défaut (1er nov → 31 mars) ──────────────────

    @Test
    fun `milieu de l'hiver (15 dec) - hors-saison`() {
        assertTrue(isOffSeason(LocalDate.of(2026, 12, 15)))
    }

    @Test
    fun `milieu de l'ete (15 juil) - en saison`() {
        assertFalse(isOffSeason(LocalDate.of(2026, 7, 15)))
    }

    @Test
    fun `1er novembre - premier jour hors-saison - true`() {
        assertTrue(isOffSeason(LocalDate.of(2026, 11, 1)))
    }

    @Test
    fun `31 octobre - veille du debut - false`() {
        assertFalse(isOffSeason(LocalDate.of(2026, 10, 31)))
    }

    @Test
    fun `31 mars - dernier jour hors-saison - true`() {
        assertTrue(isOffSeason(LocalDate.of(2026, 3, 31)))
    }

    @Test
    fun `1er avril - lendemain de la fin - false`() {
        assertFalse(isOffSeason(LocalDate.of(2026, 4, 1)))
    }

    @Test
    fun `1er janvier (la fenetre chevauche le 31 dec) - hors-saison`() {
        assertTrue(isOffSeason(LocalDate.of(2026, 1, 1)))
    }

    @Test
    fun `15 decembre - hors-saison (meme comportement en changeant d'annee)`() {
        assertTrue(isOffSeason(LocalDate.of(2025, 12, 15)))
        assertTrue(isOffSeason(LocalDate.of(2027, 12, 15)))
    }

    // ─── isOffSeason — fenêtre personnalisée ───────────────────────────────────

    private val customStart = SeasonBound(month = 6, day = 1)  // 1er juin
    private val customEnd = SeasonBound(month = 9, day = 30)   // 30 sept (fenêtre dans l'année)

    @Test
    fun `15 aout - dans la fenetre - true`() {
        assertTrue(isOffSeason(LocalDate.of(2026, 8, 15), customStart, customEnd))
    }

    @Test
    fun `1er juin - premier jour - true`() {
        assertTrue(isOffSeason(LocalDate.of(2026, 6, 1), customStart, customEnd))
    }

    @Test
    fun `31 mai - veille - false`() {
        assertFalse(isOffSeason(LocalDate.of(2026, 5, 31), customStart, customEnd))
    }

    @Test
    fun `30 sept - dernier jour - true`() {
        assertTrue(isOffSeason(LocalDate.of(2026, 9, 30), customStart, customEnd))
    }

    @Test
    fun `1er octobre - hors fenetre - false`() {
        assertFalse(isOffSeason(LocalDate.of(2026, 10, 1), customStart, customEnd))
    }

    @Test
    fun `15 dec - hors fenetre - false`() {
        assertFalse(isOffSeason(LocalDate.of(2026, 12, 15), customStart, customEnd))
    }

    // ─── nextSeasonChangeDate ──────────────────────────────────────────────────

    @Test
    fun `en hiver - prochaine date = fin hors-saison (31 mars)`() {
        val result = nextSeasonChangeDate(LocalDate.of(2026, 12, 15))
        assertEquals(OFFSEASON_END.month, result.monthValue) // 3 = mars
        assertEquals(OFFSEASON_END.day, result.dayOfMonth)   // 31
    }

    @Test
    fun `en ete - prochaine date = debut hors-saison (1er nov)`() {
        val result = nextSeasonChangeDate(LocalDate.of(2026, 7, 15))
        assertEquals(OFFSEASON_START.month, result.monthValue) // 11 = nov
        assertEquals(OFFSEASON_START.day, result.dayOfMonth)   // 1
    }

    @Test
    fun `en fevrier - fin hors-saison cette annee (mars n'est pas encore passe)`() {
        val result = nextSeasonChangeDate(LocalDate.of(2026, 2, 15))
        assertEquals(2026, result.year)
        assertEquals(3, result.monthValue) // mars
        assertEquals(31, result.dayOfMonth)
    }

    @Test
    fun `en juillet - debut hors-saison cette annee (nov pas encore passe)`() {
        val result = nextSeasonChangeDate(LocalDate.of(2026, 7, 1))
        assertEquals(2026, result.year)
        assertEquals(11, result.monthValue) // nov
    }

    // ─── isDegenerate ──────────────────────────────────────────────────────────

    @Test
    fun `start egal end - degenere`() {
        assertTrue(isDegenerate(SeasonBound(6, 15), SeasonBound(6, 15)))
    }

    @Test
    fun `start different de end (config par defaut) - non degenere`() {
        assertFalse(isDegenerate(OFFSEASON_START, OFFSEASON_END))
    }

    @Test
    fun `dates differentes - non degenere`() {
        assertFalse(isDegenerate(SeasonBound(11, 1), SeasonBound(3, 31)))
    }
}
